GVFS/GVFS.UnitTests/Tracing/QueuedPipeStringWriterTests.cs

```diff
@@ -0,0 +1,213 @@
+using System;
+using System.Collections.Generic;
+using System.IO.Pipes;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using GVFS.Common.Tracing;
+using NUnit.Framework;
+
+namespace GVFS.UnitTests.Tracing
+{
+    [TestFixture]
+    public class QueuedPipeStringWriterTests
+    {
+        [TestCase]
+        public void Stop_RaisesStateStopped()
+        {
+            QueuedPipeStringWriter writer = new QueuedPipeStringWriter(
+                () => null); // Not important since the test will never enqueue any messaged to write
+
+            // Capture event invocations
+            List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs> eventCalls = new List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs>();
+            writer.StateChanged += (_, e) => eventCalls.Add(e);
+
+            // Try to write some dummy data
+            writer.Start();
+            writer.Stop();
+
+            Assert.AreEqual(1, eventCalls.Count);
+            Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Stopped, eventCalls[0].State);
+        }
+
+        [TestCase]
+        public void MissingPipe_RaisesStateFailing()
+        {
+            const string inputMessage = "FooBar";
+            QueuedPipeStringWriter writer = new QueuedPipeStringWriter(() => throw new Exception("Failing pipe connection"));
+
+            // Capture event invocations
+            List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs> eventCalls = new List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs>();
+            writer.StateChanged += (_, e) => eventCalls.Add(e);
+
+            // Try to write some dummy data
+            writer.Start();
+            bool queueOk = writer.TryEnqueue(inputMessage);
+            writer.Stop();
+
+            Assert.IsTrue(queueOk);
+            Assert.AreEqual(2, eventCalls.Count);
+            Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Failing, eventCalls[0].State);
+            Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Stopped, eventCalls[1].State);
+        }
+
+        [TestCase]
+        public void GoodPipe_WritesDataAndRaisesStateHealthy()
```

Comment:
@jeschu1 this is the test that is failing on macOS but not in Windows.
My current theory is that the bug is in the test code and the `TestPipeWorker` or pipe-writer is not waiting for each other to be connected. Probably need to gate `TestPipeWorker::GetAllData` on the worker thread completing the train of the pipe.

Reply:
Actually.. it's because my `TestPipeWorker` needs to wait for the pipe to be drained before calling `GetAllData`.

Reply:
@mjcheetham, a fun one I missed!  So I guess it just drains faster on windows for whatever reason.

Reply:
Fixed this up by adding more signalling between the test and the test pipe worker thread, to ensure the test pipe server is actually ready to accept messages.